### ISSUE TITLE
sftp.get:  OverflowError: mode out of range  at a specific sftp server

### ISSUE THREAD

**Issue by ccwufu** [ronf/asyncssh#388]
sftp.get:  OverflowError: mode out of range  at a specific sftp server

other client works (e.g : winscp ,  sftp of cygwin) for this specific sftp server

OS : windows 10
asyncssh : 2.7.0

```python
import asyncio, asyncssh, sys

import logging

logging.basicConfig(level='DEBUG')
asyncssh.set_debug_level(3)


async def run_client():
    async with asyncssh.connect('192.168.1.1', port=8222, username='diag', password='passwd0', known_hosts=None) as conn:
        async with conn.start_sftp_client() as sftp:
            await sftp.get('CONFIG.TXT', '.')

try:
    asyncio.get_event_loop().run_until_complete(run_client())
except (OSError, asyncssh.Error) as exc:
    sys.exit('SSH connection failed: ' + str(exc))
```

```
Traceback (most recent call last):
  File "c:\tools\python396\prj0\prj0ssh1a3.py", line 17, in <module>
    asyncio.get_event_loop().run_until_complete(run_client())
  File "asyncio\base_events.py", line 642, in run_until_complete
  File "c:\tools\python396\prj0\prj0ssh1a3.py", line 13, in run_client
    await sftp.get('CONFIG.TXT', '.')
  File "c:\tools\python396\lib\site-packages\asyncssh\sftp.py", line 2264, in get
    await self._begin_copy(self, LocalFile, remotepaths, localpath, 'get',
  File "c:\tools\python396\lib\site-packages\asyncssh\sftp.py", line 2164, in _begin_copy
    await self._copy(srcfs, dstfs, srcfile, dstfile, preserve,
  File "c:\tools\python396\lib\site-packages\asyncssh\sftp.py", line 2066, in _copy
    if stat.S_ISDIR(srcattrs.permissions):
OverflowError: mode out of range
```

**details : debug level 3**
```
INFO:asyncssh:[conn=0] Auth for user diag succeeded
DEBUG:asyncssh:[conn=0, chan=0] Set write buffer limits: low-water=16384, high-water=65536
INFO:asyncssh:[conn=0, chan=0] Requesting new SSH session
DEBUG:asyncssh:[conn=0, chan=0]   Initial recv window 2097152, packet size 32768
DEBUG:asyncssh:[conn=0, pktid=10] Sent MSG_IGNORE (2), 5 bytes
  00000000: 02 00 00 00 00                                   .....
DEBUG:asyncssh:[conn=0, chan=0, pktid=11] Sent MSG_CHANNEL_OPEN (90), 24 bytes
  00000000: 5a 00 00 00 07 73 65 73 73 69 6f 6e 00 00 00 00  Z....session....
  00000010: 00 20 00 00 00 00 80 00                          . ......
DEBUG:asyncssh:[conn=0, pktid=8] Received MSG_CHANNEL_OPEN_CONFIRMATION (91), 17 bytes
  00000000: 5b 00 00 00 00 00 00 00 00 00 00 00 00 00 00 80  [...............
  00000010: 00                                               .
DEBUG:asyncssh:[conn=0, chan=0]   Initial send window 0, packet size 32768
INFO:asyncssh:[conn=0, chan=0]   Subsystem: sftp
DEBUG:asyncssh:[conn=0, pktid=12] Sent MSG_IGNORE (2), 5 bytes
  00000000: 02 00 00 00 00                                   .....
DEBUG:asyncssh:[conn=0, chan=0, pktid=13] Sent MSG_CHANNEL_REQUEST (98), 27 bytes
  00000000: 62 00 00 00 00 00 00 00 09 73 75 62 73 79 73 74  b........subsyst
  00000010: 65 6d 01 00 00 00 04 73 66 74 70                 em.....sftp
DEBUG:asyncssh:[conn=0, chan=0, pktid=9] Received MSG_CHANNEL_WINDOW_ADJUST (93), 9 bytes
  00000000: 5d 00 00 00 00 00 02 00 00                       ]........
DEBUG:asyncssh:[conn=0, chan=0] Received window adjust of 131072 bytes, new window 131072
DEBUG:asyncssh:[conn=0, chan=0, pktid=10] Received MSG_CHANNEL_SUCCESS (99), 5 bytes
  00000000: 63 00 00 00 00                                   c....
INFO:asyncssh.sftp:[conn=0, chan=0] Starting SFTP client
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending init, version=3
DEBUG:asyncssh:[conn=0, chan=0] Sending 9 data bytes
DEBUG:asyncssh:[conn=0, pktid=14] Sent MSG_IGNORE (2), 5 bytes
  00000000: 02 00 00 00 00                                   .....
DEBUG:asyncssh:[conn=0, chan=0, pktid=15] Sent MSG_CHANNEL_DATA (94), 18 bytes
  00000000: 5e 00 00 00 00 00 00 00 09 00 00 00 05 01 00 00  ^...............
  00000010: 00 03                                            ..
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sent FXP_INIT (1), 5 bytes
  00000000: 01 00 00 00 03                                   .....
DEBUG:asyncssh:[conn=0, chan=0] Reading from channel started
DEBUG:asyncssh:[conn=0, chan=0, pktid=11] Received MSG_CHANNEL_DATA (94), 18 bytes
  00000000: 5e 00 00 00 00 00 00 00 09 00 00 00 05 02 00 00  ^...............
  00000010: 00 03                                            ..
DEBUG:asyncssh:[conn=0, chan=0] Received 9 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received FXP_VERSION (2), 5 bytes
  00000000: 02 00 00 00 03                                   .....
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received version=3
DEBUG:asyncssh.sftp:[conn=0, chan=0] Adjusting for non-standard symlink implementation
INFO:asyncssh.sftp:[conn=0, chan=0] Starting SFTP get of CONFIG.TXT to .
DEBUG:asyncssh.sftp:[conn=0, chan=0] Sending lstat for CONFIG.TXT
DEBUG:asyncssh:[conn=0, chan=0] Sending 23 data bytes
DEBUG:asyncssh:[conn=0, pktid=16] Sent MSG_IGNORE (2), 5 bytes
  00000000: 02 00 00 00 00                                   .....
DEBUG:asyncssh:[conn=0, chan=0, pktid=17] Sent MSG_CHANNEL_DATA (94), 32 bytes
  00000000: 5e 00 00 00 00 00 00 00 17 00 00 00 13 07 00 00  ^...............
  00000010: 00 00 00 00 00 0a 43 4f 4e 46 49 47 2e 54 58 54  ......CONFIG.TXT
DEBUG:asyncssh.sftp:[conn=0, chan=0, pktid=0] Sent FXP_LSTAT (7), 19 bytes
  00000000: 07 00 00 00 00 00 00 00 0a 43 4f 4e 46 49 47 2e  .........CONFIG.
  00000010: 54 58 54                                         TXT
DEBUG:asyncssh:[conn=0, chan=0, pktid=12] Received MSG_CHANNEL_DATA (94), 50 bytes
  00000000: 5e 00 00 00 00 00 00 00 29 00 00 00 25 69 00 00  ^.......)...%i..
  00000010: 00 00 00 00 00 0f 00 00 00 00 00 03 ac ba 00 00  ................
  00000020: 00 01 00 00 00 01 ff ff 81 a4 60 ff e3 b9 60 ff  ..........`...`.
  00000030: e3 b9                                            ..
DEBUG:asyncssh:[conn=0, chan=0] Received 41 data bytes
DEBUG:asyncssh.sftp:[conn=0, chan=0, pktid=0] Received FXP_ATTRS (105), 37 bytes
  00000000: 69 00 00 00 00 00 00 00 0f 00 00 00 00 00 03 ac  i...............
  00000010: ba 00 00 00 01 00 00 00 01 ff ff 81 a4 60 ff e3  .............`..
  00000020: b9 60 ff e3 b9                                   .`...
DEBUG:asyncssh.sftp:[conn=0, chan=0] Received size: 240826, uid: 1, gid: 1, permissions: 37777700644, atime: Tue Jul 27 18:45:13 2021, mtime: Tue Jul 27 18:45:13 2021
**4294934948**     _**<=(sfp.py #2064 +1 : print(srcattrs.permissions))**_
DEBUG:asyncssh:[conn=0, chan=0] Sending EOF
DEBUG:asyncssh:[conn=0, pktid=18] Sent MSG_IGNORE (2), 5 bytes
  00000000: 02 00 00 00 00                                   .....
DEBUG:asyncssh:[conn=0, chan=0, pktid=19] Sent MSG_CHANNEL_EOF (96), 5 bytes
  00000000: 60 00 00 00 00                                   `....
DEBUG:asyncssh:[conn=0, chan=0, pktid=13] Received MSG_CHANNEL_REQUEST (98), 25 bytes
  00000000: 62 00 00 00 00 00 00 00 0b 65 78 69 74 2d 73 74  b........exit-st
  00000010: 61 74 75 73 00 00 00 00 00                       atus.....
INFO:asyncssh:[conn=0, chan=0] Received exit status 0
DEBUG:asyncssh:[conn=0, chan=0, pktid=14] Received MSG_CHANNEL_EOF (96), 5 bytes
  00000000: 60 00 00 00 00                                   `....
DEBUG:asyncssh:[conn=0, chan=0] Received EOF
DEBUG:asyncssh:[conn=0, chan=0, pktid=15] Received MSG_CHANNEL_CLOSE (97), 5 bytes
  00000000: 61 00 00 00 00                                   a....
INFO:asyncssh:[conn=0, chan=0] Received channel close
DEBUG:asyncssh:[conn=0, pktid=20] Sent MSG_IGNORE (2), 5 bytes
  00000000: 02 00 00 00 00                                   .....
DEBUG:asyncssh:[conn=0, chan=0, pktid=21] Sent MSG_CHANNEL_CLOSE (97), 5 bytes
  00000000: 61 00 00 00 00                                   a....
INFO:asyncssh.sftp:[conn=0, chan=0] SFTP client exited
INFO:asyncssh:[conn=0, chan=0] Closing channel
INFO:asyncssh:[conn=0, chan=0] Channel closed
INFO:asyncssh:[conn=0] Closing connection
INFO:asyncssh:[conn=0] Sending disconnect: Disconnected by application (11)
DEBUG:asyncssh:[conn=0, pktid=22] Sent MSG_IGNORE (2), 5 bytes
  00000000: 02 00 00 00 00                                   .....
DEBUG:asyncssh:[conn=0, pktid=23] Sent MSG_DISCONNECT (1), 45 bytes
  00000000: 01 00 00 00 0b 00 00 00 1b 44 69 73 63 6f 6e 6e  .........Disconn
  00000010: 65 63 74 65 64 20 62 79 20 61 70 70 6c 69 63 61  ected by applica
  00000020: 74 69 6f 6e 00 00 00 05 65 6e 2d 55 53           tion....en-US
INFO:asyncssh:[conn=0] Connection closed
Traceback (most recent call last):
  File "c:\tools\python396\prj0\prj0ssh1a3.py", line 17, in <module>
    asyncio.get_event_loop().run_until_complete(run_client())
  File "asyncio\base_events.py", line 642, in run_until_complete
  File "c:\tools\python396\prj0\prj0ssh1a3.py", line 13, in run_client
    await sftp.get('CONFIG.TXT', '.')
  File "c:\tools\python396\lib\site-packages\asyncssh\sftp.py", line 2264, in get
    await self._begin_copy(self, LocalFile, remotepaths, localpath, 'get',
  File "c:\tools\python396\lib\site-packages\asyncssh\sftp.py", line 2164, in _begin_copy
    await self._copy(srcfs, dstfs, srcfile, dstfile, preserve,
  File "c:\tools\python396\lib\site-packages\asyncssh\sftp.py", line 2066, in _copy
    if stat.S_ISDIR(srcattrs.permissions):
OverflowError: mode out of range

c:\tools\python396>
```

**Comment by ronf** [comment]
The SFTP server is definitely returning an unexpected value for "permissions" here. It appears to have all the high-order bits set in it (0xffff81a4). This triggers an incompatibility between different versions of Python based on the underlying OS's definition of mode_t, which is a 16-bit value on some OSes and 32-bit value on others. See https://bugs.python.org/issue39729 for some discussion on this.

In a quick search, I see this being reported as causing problems for other Python-based SFTP implementations as well, when they try to interpret the mode bits, particularly when talking to an AIX server. Is that the case for you?

 The SFTP spec states that this field should contain POSIX mode bits, which should only have the bottom 12 bits as non-zero. Anything above that is undefined. UNIX system define some bits above that provide additional information about file type (directory, char/block device, FIFO, symlink, socket) that fill out the rest of the bottom 16 bits, but setting anything above that can apparently lead to this error on systems with a 16-bit mode_t.

A workaround would be to mask off everything but the bottom 16 bits of the SFTP "permissions" attribute before using this value in calls to the various Python stat functions such as stat.S_ISDIR(). A good place to do this might be in SFTPAttrs.decode():

```patch
@@ -871,7 +868,7 @@ class SFTPAttrs(Record):
             attrs.gid = packet.get_uint32()

         if flags & FILEXFER_ATTR_PERMISSIONS:
-            attrs.permissions = packet.get_uint32()
+            attrs.permissions = packet.get_uint32() & 0xffff

         if flags & FILEXFER_ATTR_ACMODTIME:
             attrs.atime = packet.get_uint32()
```

Could you give this a try and see if it fixes the issue for you?

**Comment by ccwufu** [comment]
thanks.
this workaround works  at this sftp server.

BTW : the remote sftp server is not AIX server.

**Comment by ronf** [comment]
Thanks for the confirmation! This fix is now available in commit 58d776c in the "develop" branch and will be included in the next release.

**Comment by ronf** [comment]
This change is now available in AsyncSSH version 2.7.1.
